zerostash/src/commands/mount.rs

```diff
@@ -256,24 +244,102 @@ impl FilesystemMT for SimpleFs {
+        chunks.sort_by(|(a, _), (b, _)| a.cmp(b));
+        let mut stack = self.stack.lock().unwrap();
+        let stack = stack
+            .entry(real_path.to_path_buf())
+            .or_insert(RefCell::new((vec![], 0, 0)));
+        let mut stack = stack.borrow_mut();
+        let mut objectreader = self.stash.storage_reader().unwrap();
+
+        if offset > metadata_size {
             return callback(Err(libc::EINVAL));
         }
 
-        let buf: Vec<u8> = match file_content.get(real_path) {
-            Some(buf) => buf.to_vec(),
-            None => return callback(Err(libc::ENOENT)),
-        };
+        let current_read = stack.2;
+        if current_read == offset {
+            let end = usize::min(size, metadata_size - offset);
+            if stack.0.len() < end {
+                loop {
+                    let (start, chunk_p) = chunks.get(stack.1).unwrap();
+                    let arc = (metadata_size as u64, Arc::new(ChunkPointer::default()));
+                    let (next_start, _) = chunks.get(stack.1 + 1).unwrap_or(&arc);
+                    let buf_size = next_start - start;
+                    let mut buf = vec![0; buf_size as usize];
+                    objectreader.read_chunk(chunk_p, &mut buf).unwrap();
+                    stack.0.append(&mut buf);
+                    stack.1 = stack.1.wrapping_add(1);
+                    if stack.0.len() >= usize::min(size, metadata_size - offset) {
+                        break;
+                    }
+                }
+            }
+            let ret_buf = stack.0[..end].to_vec();
+            stack.0 = stack.0[end..].to_vec();
+            stack.2 = offset + end;
+            return callback(Ok(&ret_buf));
+        }
+
+        for (i, (chunk_offset, _)) in chunks.iter().enumerate() {
+            let chunk_offset = *chunk_offset as usize;
+            if offset < chunk_offset {
+                let mut chunk_index = usize::max(i - 1, 0); //0
+                let mut buf = vec![];
+                let mut from;
+                let to;
+                let (start_offset, _) = chunks.get(chunk_index).unwrap();
+                loop {
+                    let (start, chunk_p) = chunks.get(chunk_index).unwrap();
+                    let arc = (metadata_size as u64, Arc::new(ChunkPointer::default()));
+                    let (next_start, _) = chunks.get(chunk_index + 1).unwrap_or(&arc);
+                    chunk_index += 1;
+                    let buf_size = next_start - start;
+                    let mut temp_buf = vec![0; buf_size as usize];
+                    objectreader.read_chunk(chunk_p, &mut temp_buf).unwrap();
+                    buf.append(&mut temp_buf);
+                    from = offset - *start_offset as usize;
+                    if buf[from..].len() > size {
+                        from = offset - *start_offset as usize;
+                        to = from + size;
+                        break;
+                    }
```

Comment:
This seems very similar to the code chunk above. It would be great if you could extract the common functionality into a function, and if it's performance critical, mark it with `#[inline(always)]`.

It's also quite hard to tell why you need 2 loops that look so similar by reading through the code, so a few comments about your assumptions about the program state at strategic places would probably help.

Reply:
You are right, the second loop is def not needed. I got rid of that.